Single node META FATE data

kevinrr888 opened the pull request:
closes #4905 
- Moved all the fate data for a single `META` transaction into a single ZK node
	- Pushed all the data into `NodeValue` (renamed to `FateData`)
	- Previously, `FateData` stored `TStatus`, `FateReservation`, and `FateKey`. Now additionally stores the `REPO` stack and `TxInfo`.
- Status enforcement added to `MetaFateStore` (previously only possible for `UserFateStore`).
- Moved `testFateInitialConfigCorrectness()` from `UserFateStoreIT` to `UserFateIT`
- Renamed `UserFateStoreIT` to `UserFateStatusEnforcementIT` (now extends a new class `FateStatusEnforcementIT`)
	- Now only tests status enforcement (previously status enforcement + `testFateInitialConfigCorrectness()`)
- Created `MetaFateStatusEnforcementIT` (extends `FateStatusEnforcementIT`)
	- Tests that the new status enforcement in `MetaFateStore` works
- Created `FateStoreUtil`, moving the `createFateTable()` util here, created MetaFateZKSetup inner class here (the counterpart to `createFateTable()` for `UserFateStore` but sets up ZooKeeper for use in `MetaFateStore`)
- Deleted `UserFateStoreIT`s (now `UserFateStatusEnforcementIT`) method `injectStatus` replacing with the existing `setStatus` which does the same thing
- Ensured sunny day and all FATE tests still pass with these changes

I did not end up using JSON for the fate data. I was not sure what the benefit of this would be since the data needs to be stored in ZK as a byte array anyways and it seemed simpler to me to just keep the same serialization/deserialization strategy for the fate data. Maybe this could be changed to JSON in a follow on if there would be a benefit of using JSON over the current strategy.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/MetaFateStore.java:
```diff
@@ -359,29 +365,50 @@ public void setStatus(TStatus status) {
+            "Tried to delete fate data for %s when the transaction status is %s", fateId,
+            fateData.status);
+        try {
+          zk.deleteStrict(getTXPath(fateId), stat.getVersion());
```
changed to `deleteStrict` (from `recursiveDelete`) since we only expect the one node (and strict because we expect the same version). This also ensures that there is only one node as this will throw an exception if the dir is not empty

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/MetaFateStore.java:
```diff
@@ -551,30 +552,48 @@ public Stream<FateKey> list(FateKey.FateKeyType type) {
     final Optional<FateReservation> reservation;
-
-    private NodeValue(byte[] serializedData) {
+    final Deque<Repo<T>> repoDeque;
```
Thoughts about using Deque/ArrayDeque for the repo stack?

keith-turner replied:
I like it.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/MetaFateStore.java:
```diff
@@ -301,15 +282,31 @@ public void push(Repo<T> repo) throws StackOverflowException {
+              "Tried to push to the repo stack for %s when the transaction status is %s", fateId,
+              fateData.status);
+          var repoDeque = fateData.repoDeque;
+          validDequeSize.set(repoDeque.size() < MAX_REPOS);
```
StackOverflowException is an internal accumulo exception so we can easily change it.  Can change it to extend RuntimeException instead of Exception.  The could throw it from around this point in the code and that would simplify the code because could remove atomic boolean.

kevinrr888 replied:
Good suggestion. Changed StackOverflowException and fixed it's use here/avoid atomic bool.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/MetaFateStore.java:
```diff
@@ -301,15 +282,31 @@ public void push(Repo<T> repo) throws StackOverflowException {
         }
-
-        zk.putPersistentSequential(txpath + "/repo_", serialize(repo));
       } catch (StackOverflowException soe) {
```
This was preexisting code.  Not sure why this is catching and rethrowing the exception.

kevinrr888 replied:
Yeah, maybe there was more code here at some point. Deleted the catch and rethrow

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/MetaFateStore.java:
```diff
@@ -335,18 +342,17 @@ public void setStatus(TStatus status) {
       try {
-        zk.mutateExisting(getTXPath(fateId), currSerializedData -> {
-          NodeValue currNodeVal = new NodeValue(currSerializedData);
+        zk.mutateExisting(getTXPath(fateId), currSerFateData -> {
```
Not sure if this would be effective, feel free to ignore this comment.  Was wondering if an internal util function like the following would reduce boiler place code a bit.  Can focus on transforming the FateData object.  Noticed a pattern that seems the same with this setStatus function, but not sure if its always the same.

```java
    private void mutate(FateId fateId, UnaryOperator<FateData<T>> fateDataOp) {
      try {
      zk.mutateExisting(getTXPath(fateId), currSerFateData -> {
        FateData<T> currFateData = new FateData<>(currSerFateData);
        FateData<T> newFateData = fateDataOp.apply(currFateData);
        if(newFateData == null){
          return null;
        }else{
          return newFateData.serialize();
        }

      });
      } catch (KeeperException | InterruptedException | AcceptableThriftTableOperationException e) {
        throw new IllegalStateException(e);
      }
    }
```

kevinrr888 replied:
Thanks, I added the suggested method. I think this cleans up the code and makes things a bit more readable and reusuable.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/meta/MetaFateIT.java:
```diff
@@ -19,67 +19,55 @@
-    zk = szk.getZooReaderWriter();
-    zk.mkdirs(ZK_ROOT + Constants.ZFATE);
-    zk.mkdirs(ZK_ROOT + Constants.ZTABLE_LOCKS);
+    FateStoreUtil.MetaFateZKSetup.setup(tempDir);
```
Consolidating duplicate code in the test is nice.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/meta/MetaMultipleStoresIT.java:
```diff
@@ -21,35 +21,28 @@
   public static void setup() throws Exception {
-    SZK = new ZooKeeperTestingServer(TEMP_DIR);
-    ZK = SZK.getZooReaderWriter();
-    FATE_DIR = Constants.ZFATE;
```
It may have been ok storing everything in ZK directly at /fate

kevinrr888 replied:
yeah, that too

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/FateStatusEnforcementIT.java:
```diff
@@ -0,0 +1,98 @@
+    testOperationWithStatuses(() -> {
+      // Setup for pop: Ensure there something to pop by first pushing
+      try {
+        txStore.setStatus(ReadOnlyFateStore.TStatus.NEW);
```
Not sure, could have been done that way because the method was available.  Or maybe the code evolved over time and it used to make sense.

keith-turner reviewed (commented):
These changes look good.  The test cleanup and reorg is nice. The only possible problem I saw was the repo ser/deser.  Wondering if the `!` char could occur in serialized data.  Would be nice to avoid that.

kevinrr888 commented:
Addressed in 29edaa2cd79ab2920edc9d2a4355e9f761ff78d8:
- Got rid of END_REPO_DATA marker, replacing with simply writing the number of repos to read
- Avoid AtomicBoolean in `MetaFateStore.FateTxStoreImpl.push(repo)` by changing StackOverflowException to instead be a RuntimeException (previously Exception)
- Deleted unnecessary preexisting catch and immediate re-throw of a StackOverflowException in `MetaFateStore.FateTxStoreImpl.push(repo)`
- Cleaned up and refactored `MetaFateStore` methods which mutate existing FateData; now reuse same pattern across these methods: all call new method `MetaFateStore.mutate()`

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/MetaFateStore.java:
```diff
@@ -596,24 +615,103 @@ private Optional<FateReservation> deserializeFateReservation(DataInputBuffer buf
+        if (!repoDeque.isEmpty()) {
+          byte[] serializedRepo;
+          // iterates from top/first/head to bottom/last/tail
+          for (Repo<T> repo : repoDeque) {
```
Yeah, I also don't like the arbitrary char to indicate the end. Changed.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/MetaFateStore.java:
```diff
@@ -202,25 +213,27 @@ public void deleteDeadReservations() {
+      } catch (KeeperException.NoNodeException nne) {
         // the node has since been deleted. Can safely ignore
-      } catch (KeeperException | InterruptedException | AcceptableThriftTableOperationException e) {
+      } catch (KeeperException e) {
         throw new RuntimeException(e);
       }
```
With new structure with new `mutate()` method, this will only throw `RuntimeException` if it is a `KeeperException` and `IllegalStateException` otherwise... Previously was throwing RuntimeException for all cases... not sure if it matters to keep the RuntimeException for InterruptedException or AcceptableThriftTableOperationException, but thought I'd mention it in case it does

keith-turner replied:
This seems ok

kevinrr888 commented:
Resolved the conflicts from the recent seed transaction speedup. Thankfully, those changes were mostly for UserFateStore and these are for MetaFateStore, so conflicts were very easy
